virtdbg/vmx/vmm.c

```diff
@@ -0,0 +1,290 @@
+#include <vmx/vmm.h>
+#include <mm/pmm.h>
+#include <arch/intrin.h>
+#include <stdint.h>
+#include <vmx/ept.h>
+#include <stddef.h>
+#include <util/except.h>
+#include <virtdbg.h>
+
+extern void vm_resume(guest_state_t *t);
+extern ept_entry_t* g_root_pa;
+
+//enable vmx operation
+err_t vmxon() {
+    err_t err = NO_ERROR;
+    uint32_t* vmxon_region = pallocz_aligned(0x1000, 0x1000);
+
+    size_t control = __rdmsr(0x3a);
+    if((control & (0x1 | 0x4)) != (0x1 | 0x04)) {
+        //Enabled outside of SMX and lock bit.
+        __wrmsr(0x3a, control | 0x1 | 0x4);
+    }
+
+    uint64_t cr0 = __readcr0().raw;
+    cr0 &= __rdmsr(0x487);
+    cr0 |= __rdmsr(0x486);
+    __writecr0((ia32_cr0_t) { .raw = cr0 });
+
+    ia32_cr4_t cr4 = __readcr4();
+    cr4.raw |= 1 << 13;
+    cr4.raw &= __rdmsr(0x489);
+    cr4.raw |= __rdmsr(0x488);
+    __writecr4(cr4);
+
+    //Set vmx revision
+    uint32_t vmx_revision = __rdmsr(0x480);
+    *vmxon_region = vmx_revision;
+    uint16_t successful = 0;
+    asm volatile(
+        "vmxon %1;"
+        "movq $0, %%rdx;"
+        "success:"
+        "movq $1, %%rdx;"
+        :"=d"(successful)
+        :"m"(vmxon_region)
+        :"memory", "cc"
+    );
+
+    CHECK_ERROR(successful, ERROR_UNSUPPORTED, "VMX not supported");
+
+cleanup:
+    return err;
+}
+
+static inline int vmptrld(uintptr_t vmcs) {
+    uint8_t ret;
+    asm volatile (
+        "vmptrld %[pa];"
+        "setna %[ret];"
+        : [ret]"=rm"(ret)
+        : [pa]"m"(vmcs)
+        : "cc", "memory");
+    return ret;
+}
+
+static inline int vmclear(uintptr_t vmcs) {
```

Comment:
either assert on these bits or `CHECK` them for the `vmclear` and `vmwrite` as well